### PR TITLE
fix(select): ensure select list always has correct height to fit content

### LINKED ISSUE #5341: Select component cuts off option text when it wraps onto a new line
### Current behaviour

Select component cuts off option text when it wraps onto a new line

image

### Expected behaviour

Select component should display option text when it wraps onto a new line without cutting it off.

### CodeSandbox or Storybook URL

https://codesandbox.io/s/select-options-height-resize-uy2shp?file=/src/index.js

### JIRA Ticket (Sage Only)

SBS-38045

### Suggested Solution

_No response_

### Carbon Version

109.1.2

### Design Tokens Version

2.40.0

### What browsers are you seeing the problem on?

Chrome

### What Operating System are you seeing the problem on?

MacOS

### Anything else we should know?

_No response_

### Confidentiality

- [X] I confirm there is no confidential or commercially sensitive information included.

### PR DESCRIPTION
**Pull request by robinzigmond**
Use a callback ref to ensure that the the code that calculates the list height is ran whenever the
DOM node is attached.

fix #5341

### Proposed behaviour

Select list should always be the correct height to show all its options, with the correct padding/margin, no matter how many times the text wraps.

### Current behaviour

In some circumstances (needs few enough options for there not to be a scrollbar), options with long text that wraps cause the list height to be too small and some parts of some options therefore are not visible.

### Checklist

- [X] Commits follow our style guide
- [X] Related issues linked in commit messages if required
- [X] Screenshots are included in the PR if useful
- [X] All themes are supported if required
- [X] Unit tests added or updated if required
- [X] Cypress automation tests added or updated if required
- [X] Storybook added or updated if required
- [X] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [X] Typescript `d.ts` file added or updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Additional context

There's a bit of history behind this bug/change that reviewers may find useful.

This was first reported as issue #4489.
Fixed in [this PR](https://github.com/Sage/carbon/pull/4573) by adding `listRef.current` to the `useLayoutEffect`'s dependency array - this works because the height needs to be recalculated when the select list container appears in the DOM, which wasn't happening previously.
This fix was removed in [this PR](https://github.com/Sage/carbon/pull/4949), as part of a much wider set of changes, it seems without any comment at all!

While restoring the dependency fixes the problem, I assume the reason for the removal was that it causes lint warnings saying that a `ref.current` shouldn't be used as an effect dependency - effects only run after a render but an update to a ref does not cause a rerender to happen, so this only works when we are fortunate enough for a prop or state change to happen at about the same time, and I didn't want to rely on this.

It seems the way to guarantee code to run when a ref is updated is to use a callback ref, so that is what I have implemented here.

### Testing instructions

The following CodeSandbox is an example of the broken behaviour.
You can see the new behaviour by looking at the version in the comment by `codesandbox[bot]`.

Codesandbox from original issue: https://codesandbox.io/s/select-options-height-resize-uy2shp?file=/src/index.js

I have changed the "Select/Test" story to a configuration that is more likely to trigger the bug, so this can also be tested (and I have added a new Cypress test to hopefully catch any regressions in this behaviour).

### CODE REVIEW

File: src/components/select/select-list/select-list.component.js [hunk 1]
```diff
@@ -318,17 +335,7 @@ const SelectList = React.forwardRef(
...
-
-      newHeight = listRef.current.clientHeight;
-
-      if (listActionButtonRef.current) {
-        newHeight += listActionButtonRef.current.parentElement.clientHeight;
-      }
-
-      setListHeight(`${newHeight}px`);
-    }, [children]);
+    useLayoutEffect(updateListHeight, [children]);
```
Comment (mkrds) [inline src/components/select/select-list/select-list.component.js:338]: I might be wrong but I think we don't need `useLayoutEffect` anymore here.

Comment (robinzigmond) [reply src/components/select/select-list/select-list.component.js:338]: OK, I'll give it a go. I never gave it much thought, I just assumed as it was already using `useLayoutEffect` there would be a good reason (particularly as it's setting the size of DOM elements).

**Review by DipperTheDan — COMMENTED** [review]
I'm happy with the fix. Just address the comment left by @mkrds and I'll happily approve the PR.

**Review by DipperTheDan — DISMISSED** [review]

**Review by mkrds — DISMISSED** [review]

File: src/components/select/simple-select/simple-select.test.js [hunk 2]
```diff
@@ -996,4 +1011,14 @@ context("Tests for Simple Select component", () => {
...
     );
   });
+
+  describe("check height of Select list when opened", () => {
+    it("with long option text, should not cut off any text", () => {
+      CypressMountWithProviders(<SimpleSelectWithLongWrappingTextComponent />);
+
+      selectText().click();
+      selectListWrapper().should("be.visible");
+      selectListWrapper().should("have.css", "height", "152px");
```
Comment (divyajindel) [inline src/components/select/simple-select/simple-select.test.js:1021]: describe("check height of Select list when opened", () => {
    it("should not cut off any text with long option text", () => {
      CypressMountWithProviders(<SimpleSelectWithLongWrappingTextComponent />);
      selectText().click();
      selectListWrapper().should("have.css", "height", "152px").and("be.visible");
    });
  });

**Review by divyajindel — CHANGES_REQUESTED** [review]

**Review by divyajindel — APPROVED** [review]

**Pull request merged**